check-meta: add a teams attribute

RossComputerGuy opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

This is something I've had in mind for some time. It'd be great for `meta` to be aware of teams so if a package is only owned by a team or set of teams, the maintainers could be pulled from the members list of each team. This could go onto the package search and make it possible to see which team owns a package.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

RossComputerGuy committed:
check-meta: add a teams attribute

phaer commented:
We had teams owning packages since at least 2019, see https://github.com/NixOS/nixpkgs/blob/master/maintainers/team-list.nix

Could you elaborate what the benefits/differences of a dedicated attribute would be from your perspective?

SigmaSquadron commented:
Forgive me if I'm missing something obvious, but I thought teams were just a fancy way to make a list of maintainers. Every package seems to use `lib.teams.team.members`, which is a list of attrs, and is thus accepted by the `check-meta` in HEAD.

If we started adding the top-level team attribute instead of `team.members`, (no package currently does this, to my knowledge) we'd probably break every consumer of `meta.maintainers`, which includes the package search, `nix-owners`, and probably some more software.

SigmaSquadron reviewed (approved):
Nevermind, I did miss something obvious, sorry! This seems like a good change. I'm wondering if we can also add the members of a team to `meta.maintainers` automatically so team-maintained packages don't appear as unmaintained to consumers of `meta.maintainers` in a follow-up PR.

RossComputerGuy commented:
That's all good heh. Yeah, this is going to be rolled out slowly. Once this lands, there'll be communication to the consumers of `meta.maintainers` to look into using it. I believe we already have that done via `hasNoMaintainers` covering teams (visible in the diff of this PR).

RossComputerGuy committed:
docs: add new meta.teams

SigmaSquadron commented on doc/stdenv/meta.chapter.md:
```diff
@@ -91,6 +91,10 @@ For details, see [Source provenance](#sec-meta-sourceProvenance).
 
+### `teams` {#var-meta-teams}
+
+A list of the teams of this Nix expression. Teams are defined in [`nixpkgs/maintainers/team-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/maintainers/team-list.nix).
```
```suggestion
A list of the teams of this Nix expression. Teams are defined in [`nixpkgs/maintainers/team-list.nix`](https://github.com/NixOS/nixpkgs/blob/master/maintainers/team-list.nix), and can be defined in a package with `meta.teams = with lib.teams; [ team1 team2 ]`.
```
Similar example to `maintainers` above.

RossComputerGuy replied:
Done, CI should also be fixed too.

philiptaron commented:
I must be dense -- I'm suffering from the same "why?" question as @SigmaSquadron had initially. What's the difference, in your mind, between a "maintainer", a "team", and a team that's mentioned in the maintainers' list?

RossComputerGuy commented:
The `maintainers` attribute imo loosely defines the team but `teams` defines a more coordinated group. This gives consumers of `meta` more information on who owns the package. This allows us to find packages based on whether they're "critical" for release based on the release ping attribute in the team. It's also possible to ping just the team rather than the maintainers. `maintainers = team.members` doesn't communicate the actual team because `team` isn't visible, only the members.

numinit commented:
I see the purpose, this way we can set up pings for teams, not just all members on that team. I think it would be more useful to have it be backwards compatible, though -- if you start specifying `teams` it just does the same thing as adding all those members as maintainers under the hood, for now, until we find better ways to expose that meta somewhere.

RossComputerGuy commented:
Hmm yeah, we can do that for 25.05 and possibly change it in 25.11.

numinit commented:
It would definitely help some of this annoying duplication: https://github.com/NixOS/nixpkgs/pull/396869/files

RossComputerGuy committed:
ci/eval/compare: add meta.teams

numinit committed:
maintainers/scripts: add get-maintainer-pings-between.sh

Sometimes it is useful to see which maintainers your change will ping,
to know who you are going to spam in advance.

numinit committed:
ci/eval/compare: support optional byName argument

Sometimes it is quite useful to output names instead of GitHub IDs, e.g.
for maintainer scripts that show you who you would ping. Add this as an
option, but keep the existing default.

numinit commented on pkgs/stdenv/generic/check-meta.nix:
```diff
@@ -108,7 +108,10 @@ let
+  hasNoMaintainers =
+    attrs:
+    (attrs ? meta.maintainers && (length attrs.meta.maintainers) == 0)
+    || (attrs ? meta.teams && (length attrs.meta.teams) == 0);
```
This should be &&

RossComputerGuy replied:
Done

numinit commented:
Some pair programming in Carlsbad for this one. :-)

numinit commented on maintainers/scripts/get-maintainer-pings-between.sh:
```diff
@@ -0,0 +1,78 @@
+    nix-build ci -A eval.singleSystem \
+        --arg evalSystem "$system" \
+        --arg attrpathFile "$workdir/paths/paths.json" \
+        --arg chunkSize 10000 \
```
Probably should make this configurable via environment variable. Since two evals are possibly happening at once, it may use twice the RAM.

RossComputerGuy replied:
Done

numinit commented on maintainers/scripts/get-maintainer-pings-between.sh:
```diff
@@ -0,0 +1,78 @@
+git -C "$repo" diff --name-only "$rev1" "$rev2" \
+    | jq --raw-input --slurp 'split("\n")[:-1]' | tee "$TMPDIR/touched-files.json" >&2
+
+# Runs an eval in the given worktree, outputting the path to $TMPDIR/$1.path.
```
Probably the latter part of this comment is now inaccurate.

RossComputerGuy replied:
Done

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-ready-for-review/3032/5394

numinit commented:
Manual build looks borked for some reason?

RossComputerGuy commented:
@numinit Been mentioned in the PR which broke it https://github.com/NixOS/nixpkgs/pull/369283#issuecomment-2813510460

numinit commented:
Aha, ty.

numinit reviewed (approved):
Looks good, but I'm not a CI codeowner. Though this certainly taught me more than I ever wanted to know about it!

winterqt commented:
Supportive of reverting this before branch-off.

numinit commented:
I'm fine with whatever is decided: ultimately, any given PR is less important than its overall place in the project. If this (as well as the treewide) were the wrong move, let's talk about it.

However, I will point out that the PR was up for 21 days prior to merging for exactly that purpose: to gather feedback. If this was poorly conceived, the past month would have been the correct time to deliver that feedback, rather than waiting until the thirteenth hour and calling for a revert.

winterqt commented:
> However, I will point out that the PR was up for 21 days prior to merging for exactly that purpose: to gather feedback. If this was poorly conceived, the past month would have been the correct time to deliver that feedback, rather than waiting until the thirteenth hour and calling for a revert.

Do note that due to GitHub limitations, a good chunk of the codeowners unfortunately were not pinged, and as such really can't be blamed for missing it.

RossComputerGuy commented:
> Do note that due to GitHub limitations, a good chunk of the codeowners unfortunately were not pinged, and as such really can't be blamed for missing it.

That is more of with #400458 rather than this PR. I did mention this PR several times in the NixOS Stdenv Matrix chat, though nobody objected to it. I did see that over 14 people saw it on that chat several days before this merged so it's not like this was going to happen without prior knowledge.

winterqt commented:
It applies to @mweinelt specifically, though, as `ci/` is owned by the security team.

RossComputerGuy commented:
Huh, though the bot could've pinged 5 more people right? There's a cap of 15 reviewers, at least through the web UI, and there was 10 people in total pinged. At least philiptaron was pinged so that brings down the number of people who needed to be pinged. Not sure why touching `ci/eval` didn't ping more people though.

numinit commented:
@winterqt That is a good point, and strange that didn't happen. I'd have thought codeowner pings would have handled it, but alas. ci is listed in the owners file, so I was under the assumption that the right people had been pinged. ☹️

That aside, I am up for doing whatever we think is best as a project. This change was meant to reduce boilerplate and ultimately ping @NixOS/* down the line. If we didn't pull in the right people, well, we should just make sure everyone is aware next time.

RossComputerGuy commented:
I did notice a PR that also should've pinged codeowners, #400408, but it didn't. Not sure what happened there, but it looks like maybe the ping bot or GitHub itself is flaky with pings?

numinit commented:
I think I may have figured it out: see [OWNERS](https://github.com/NixOS/nixpkgs/blob/master/ci/OWNERS#L22) and [the docs](https://docs.github.com/en/repositories/managing-your-repositorys-settings-and-features/customizing-your-repository/about-code-owners). It's subtle: ci was listed without a trailing slash, so it doesn't apply recursively. @NixOS/stdenv was pinged since /pkgs/stdenv/generic/check-meta.nix was touched, that's why @philiptaron got pinged, but changing something in a subdirectory of ci didn't ping the people on the list for ci.

> In this example, @doctocat owns any files in the build/logs directory at the root of the repository and any of its subdirectories.

Note the difference between this and:

> In this example, @octocat owns any file in a `/logs` directory such as `/build/logs`, `/scripts/logs`, and `/deeply/nested/logs`.

So it may be a matter of adding a trailing / or a `/**`.

winterqt commented:
Oh, [what lovely timing](https://github.com/NixOS/nixpkgs/pull/405153). lol.

> What's the state here, will this be kept or reverted?

See #402991, the change is effectively unnoticeable now, as `meta.maintainers` now contains `meta.teams` entries.

SigmaSquadron commented:
Seems like that check was already broken, and this simply brought it to light by making the check more explicitly broken.
